src/stargate.rs

```diff
@@ -0,0 +1,106 @@
+use crate::error::AnyResult;
+use crate::{AppResponse, CosmosRouter};
+use anyhow::bail;
+use cosmwasm_std::{Addr, Api, Binary, BlockInfo, CustomQuery, Querier, Storage};
+use schemars::JsonSchema;
+use serde::de::DeserializeOwned;
+use std::fmt::Debug;
+
+/// Stargate interface.
+///
+/// This trait provides the default behaviour for all functions
+/// that is equal to [StargateFailing] implementation.
+pub trait Stargate {
+    /// Processes stargate messages.
+    ///
+    /// The `CosmosMsg::Stargate` message is unwrapped before processing.
+    /// The `type_url` and `value` attributes of `CosmosMsg::Stargate`
+    /// are passed directly to this handler.
+    fn execute<ExecC, QueryC>(
+        &self,
+        api: &dyn Api,
+        storage: &mut dyn Storage,
+        router: &dyn CosmosRouter<ExecC = ExecC, QueryC = QueryC>,
+        block: &BlockInfo,
+        sender: Addr,
+        type_url: String,
+        value: Binary,
+    ) -> AnyResult<AppResponse>
+    where
+        ExecC: Debug + Clone + PartialEq + JsonSchema + DeserializeOwned + 'static,
+        QueryC: CustomQuery + DeserializeOwned + 'static,
+    {
+        let _ = (api, storage, router, block);
+        bail!(
+            "Unexpected stargate message: (type_ur = {}, value = {:?}) from {:?}",
+            type_url,
+            value,
+            sender
+        )
```

Comment:
Usually, I'm not a fan of providing failing default implementations, but I think in this case it makes sense, since most users probably only want to implement either query or execute